clang/lib/Driver/ToolChains/SYCL.cpp

```diff
@@ -476,15 +476,38 @@ static void addImpliedArgs(const llvm::Triple &Triple,
                            llvm::opt::ArgStringList &CmdArgs) {
   // Current implied args are for debug information and disabling of
   // optimizations.
-  // TODO: Add support for other architectures (gen, x86_64) as those are
-  // being defined.
-  if (Triple.getSubArch() == llvm::Triple::NoSubArch ||
-      Triple.getSubArch() == llvm::Triple::SPIRSubArch_fpga) {
+  auto addArgs = [&](llvm::opt::ArgStringList &CurArgList, bool IsGen = false) {
     if (Arg *A = Args.getLastArg(options::OPT_g_Group, options::OPT__SLASH_Z7))
       if (!A->getOption().matches(options::OPT_g0))
-        CmdArgs.push_back("-g");
+        CurArgList.push_back("-g");
     if (Args.getLastArg(options::OPT_O0))
-      CmdArgs.push_back("-cl-opt-disable");
+      CurArgList.push_back(IsGen ? "-O0" : "-cl-opt-disable");
+  };
+  // For FPGA and default device, pass along -g -cl-opt-disable
+  if (Triple.getSubArch() == llvm::Triple::NoSubArch ||
+      Triple.getSubArch() == llvm::Triple::SPIRSubArch_fpga) {
+    addArgs(CmdArgs);
+    return;
+  }
+  bool IsGen = Triple.getSubArch() == llvm::Triple::SPIRSubArch_gen;
+  // For CPU/Gen pass option list
+  if (Triple.getSubArch() == llvm::Triple::SPIRSubArch_x86_64 || IsGen) {
+    llvm::opt::ArgStringList BeArgs;
+    addArgs(BeArgs, IsGen);
+    if (!BeArgs.empty()) {
+      SmallString<128> BeOpt;
+      if (IsGen)
+        CmdArgs.push_back("-options");
+      else
+        BeOpt = "--bo=";
+      for (unsigned I = 0; I < BeArgs.size(); ++I) {
+        if (I)
+          BeOpt += ' ';
+        BeOpt += BeArgs[I];
+      }
+      CmdArgs.push_back(Args.MakeArgString(BeOpt));
```

Comment:
I believe this whole block could be reduced by reworking `addArgs()` into accepting the offline compiler option name. Aiming to achieve:
```
if (Triple.getSubArch() == llvm::Triple::SPIRSubArch_gen)
  addArgs(CmdArgs, IsGen, "-options");
else if (Triple.getSubArch() == llvm::Triple::SPIRSubArch_x86_64)
  addArgs(CmdArgs, IsGen, "--bo");
// Similar adjustments for SPIR-V and FPGA, passing `""` as the last argument or smth
```

Reply:
I think reworking to add the specific arguments used for the various AOT tool compilation command lines adds a bit too much complexity due to how the option arguments are handled differently.  I would prefer to leave this as is if that is OK.

Reply:
I can't really tell if the handling is much different - looks like both `--bo` and `-options` are handled by the same LOC 504-508:
```
for (unsigned I = 0; I < BeArgs.size(); ++I) {
   if (I)
     BeOpt += ' ';
   BeOpt += BeArgs[I];
}
```
Therefore, the suggestion was merely aimed at moving this into the lambda and hand-passing the `BeOpt` argument instead of calculating it based on `IsGen`, which may look kinda convoluted.

If the above is still missing some nuances/an undesirable approach, let's resolve just this and go ahead with merging :)

Reply:
Yes, that behavior is handled with both options, but how it is put together is different.  The nuance that makes it odd to work with is how the `CmdArgs` is populated.  Now that I think about it more, I may be able to pull it in.  I'll get back to you.

Reply:
Updated the behavior.  Pulling in the additional information (option name, etc) into the lambda clarified the code so that the lambda was no longer needed.